src/wp-includes/class-wp-theme-json.php

```diff
@@ -0,0 +1,374 @@
+<?php
+/**
+ * Process of structures that adhere to the theme.json schema.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Class that encapsulates the processing of
+ * structures that adhere to the theme.json spec.
+ */
+class WP_Theme_JSON {
+
+	/**
+	 * Container of data in theme.json format.
+	 *
+	 * @var array
+	 */
+	private $theme_json = null;
+
+	const VALID_TOP_LEVEL_KEYS = array(
+		'version',
+		'settings',
+	);
+
+	const ALLOWED_SETTINGS = array(
+		'color'      => array(
+			'custom'         => null,
+			'customGradient' => null,
+			'duotone'        => null,
+			'gradients'      => null,
+			'link'           => null,
+			'palette'        => null,
+		),
+		'custom'     => null,
+		'layout'     => null,
+		'spacing'    => array(
+			'customMargin'  => null,
+			'customPadding' => null,
+			'units'         => null,
+		),
+		'typography' => array(
+			'customFontSize'        => null,
+			'customLineHeight'      => null,
+			'dropCap'               => null,
+			'fontSizes'             => null,
+		),
+	);
+
+	const LATEST_SCHEMA = 1;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param array $theme_json A structure that follows the theme.json schema.
+	 */
+	public function __construct( $theme_json = array() ) {
+		if ( ! isset( $theme_json['version'] ) || self::LATEST_SCHEMA !== $theme_json['version'] ) {
+			$this->theme_json = array();
+			return;
+		}
+
+		$valid_block_names = array_keys( WP_Block_Type_Registry::get_instance()->get_all_registered() );
+		$this->theme_json  = self::sanitize( $theme_json, $valid_block_names );
+	}
+
+	/**
+	 * Returns the allowed settings in a theme.json structure.
+	 *
+	 * @return array
+	 */
+	private static function get_allowed_settings() {
+		return apply_filters( 'theme_json_allowed_settings', self::ALLOWED_SETTINGS );
```

Comment:
An alternative to this would be to allow filtering the whole schema at once. Styles are settings are many times inter-related (but not dependant) on each other: see for example `settings.border` and `styles.border`. There may also be a need to filter the allowed top-level keys.

Reply:
Ok, I've done this at https://github.com/WordPress/wordpress-develop/pull/1262/commits/4691d19a0ee037ae3573142d486a57b3bdadb4ff